### PR TITLE
SC-2078: Audit history indexer flink job

### PR DESCRIPTION
**Pull request by kumarks1122**

### CODE REVIEW

File: audit-history-indexer/src/main/resources/audit-history-indexer.conf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  consumer.parallelism = 1
+  parallelism = 1
+  window.time = 60
+}
+
+redis {
+  database {
+    relationCache.id = 10
+    collectionCache.id = 5
+  }
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/resources/audit-history-indexer.conf:18]: this configuration is not required.

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.{Calendar, Date}
+import scala.collection.mutable.Map
+import scala.collection.immutable
+
+trait AuditHistoryIndexerService {
+  private[this] lazy val logger = LoggerFactory.getLogger(classOf[AuditHistoryIndexerService])
+
+  def processEvent(event: Event, metrics: Metrics)(implicit esUtil: ElasticSearchUtil, config: AuditHistoryIndexerConfig): Unit = {
+    if (event.isValid) {
+      logger.debug("Audit learning event received")
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:19]: change to info log and mention for which identifier the audit is creating for.

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      try {
+        val record = getAuditHistory(event)
+        val identifier = event.nodeUniqueId
+        logger.info("Audit record created for " + identifier)
+        val document = JSONUtil.serialize(record)
+        logger.debug("Saving the record into ES")
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:25]: remove this
```suggestion

```

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+trait AuditHistoryIndexerService {
+  private[this] lazy val logger = LoggerFactory.getLogger(classOf[AuditHistoryIndexerService])
+
+  def processEvent(event: Event, metrics: Metrics)(implicit esUtil: ElasticSearchUtil, config: AuditHistoryIndexerConfig): Unit = {
+    if (event.isValid) {
+      logger.debug("Audit learning event received")
+      try {
+        val record = getAuditHistory(event)
+        val identifier = event.nodeUniqueId
+        logger.info("Audit record created for " + identifier)
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:23]: move this to last statement.

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        val document = JSONUtil.serialize(record)
+        logger.debug("Saving the record into ES")
+        val indexName = getIndexName(event.ets)
+        esUtil.addDocumentWithIndex(document, indexName)
+        metrics.incCounter(config.successEventCount)
+      } catch {
+        case ex: Exception =>
+          logger.error("Error while processing message :: " + event.getJson + " :: ", ex)
+          metrics.incCounter(config.failedEventCount)
+      }
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:33]: What are the possible scenarios to get the Exception here. If it is a ES connection issue, the job should fail. So, throw the exception to next level.

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    else logger.info("Learning event not qualified for audit")
+  }
+
+  private def getIndexName(ets: Long)(implicit config: AuditHistoryIndexerConfig):String = {
+    val cal = Calendar.getInstance
+    cal.setTime(new Date(ets))
+    config.auditHistoryIndex + "_" + cal.get(Calendar.YEAR) + "_" + cal.get(Calendar.WEEK_OF_YEAR)
+  }
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:42]: We should consider the timezone also for creating the name of the index. Did you use the code from Samza for this?

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+  def getAuditHistory(transactionDataMap: Event): AuditHistoryRecord = {
+    var nodeUniqueId = transactionDataMap.nodeUniqueId
+    if (StringUtils.endsWith(nodeUniqueId, ".img")) {
+      nodeUniqueId = StringUtils.replace(nodeUniqueId, ".img", "")
+    }
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:48]: Why the if condition. Always replace it.

File: audit-history-indexer/src/main/scala/org/sunbird/job/task/AuditHistoryIndexerConfig.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  // ES Configs
+  val esConnectionInfo = config.getString("es.basePath")
+
+  var auditHistoryIndex = "kp_audit_log"
+  val operationCreate = "CREATE"
+  val auditHistoryIndexType = "ah"
+
+  // Redis Configurations
+  val relationCacheStore: Int = config.getInt("redis.database.relationCache.id")
+  val collectionCacheStore: Int = config.getInt("redis.database.collectionCache.id")
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/task/AuditHistoryIndexerConfig.scala:42]: Do we need this?

File: audit-history-indexer/src/test/resources/test.conf [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  consumer.parallelism = 1
+  parallelism = 1
+  window.time = 60
+}
+
+redis {
+  database {
+    relationCache.id = 10
+    collectionCache.id = 5
+  }
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/test/resources/test.conf:18]:
```suggestion

```

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    val summary = setSummaryData(transactionDataMap)
+    AuditHistoryRecord(nodeUniqueId, transactionDataMap.objectType, transactionDataMap.label, transactionDataMap.graphId, transactionDataMap.userId, transactionDataMap.requestId, JSONUtil.serialize(transactionDataMap.transactionData), transactionDataMap.operationType, transactionDataMap.createdOnDate, summary)
+  }
+
+  private def setSummaryData(transactionDataMap: Event): String = {
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:53]: Is this summaryData required?

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: audit-history-indexer/project/build.properties [hunk 11]
```diff
@@ -0,0 +1 @@
+sbt.version=1.4.3
```
Comment (maheshkumargangula) [inline audit-history-indexer/project/build.properties:1]: what is this file?

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala (hunk 10 above)
Comment (kumarks1122) [reply audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:54]: Yes @maheshkumargangula .

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+          list = entryVal.asInstanceOf[List[AnyRef]]
+          if (null != list && list.nonEmpty) tags ++= Map("addedTags"-> list.size)
+          else tags ++= Map("addedTags"-> 0)
+          summaryData ++= Map("tags"-> tags)
+
+        case "removedTags" =>
+          list = entryVal.asInstanceOf[List[AnyRef]]
+          if (null != list && list.nonEmpty) tags ++= Map("removedTags"-> list.size)
+          else tags ++= Map("removedTags"-> 0)
+          summaryData ++= Map("tags"-> tags)
```
Comment (maheshkumargangula) [inline audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:88]: remove this.

**Review by maheshkumargangula — APPROVED** [review]

File: audit-history-indexer/project/build.properties (hunk 11 above)
Comment (kumarks1122) [reply audit-history-indexer/project/build.properties]: @maheshkumargangula : SBT build property file was added during new job creation. It is removed now.

File: audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala (hunk 10 above)
Comment (maheshkumargangula) [reply audit-history-indexer/src/main/scala/org/sunbird/job/service/AuditHistoryIndexerService.scala:54]: remove summary computation.

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**